module/executiondatasync/optimistic_sync/execution_result/info_provider.go

```diff
@@ -94,6 +103,19 @@ func (e *Provider) ExecutionResultInfo(
 		return nil, fmt.Errorf("failed to choose execution nodes for block ID %v: %w", blockID, err)
 	}
 
+	sealedHeader, err := e.state.Sealed().Head()
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup sealed header: %w", err)
+	}
+	header, err := e.headers.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header by block ID %v: %w", blockID, err)
+	}
+	// If block is sealed and criteria cannot be met return an error
+	if header.Height <= sealedHeader.Height && len(subsetENs) < len(criteria.RequiredExecutors) {
+		return nil, optimistic_sync.NewCriteriaNotMetError(blockID)
+	}
```

Comment:
this check assumes that the provided block is finalized, but I don't think that's enforced anywhere. this requires an additional lookup of `BlockIDByHeight` to determine if the finalized blockID matches the provided blockID. if it is, the following check is OK. if it's not, then the requested block conflicts with a finalized (or sealed) block, which should also return a sentinel error.